awscli/customizations/cloudformation/modules/process.py

```diff
@@ -0,0 +1,1306 @@
+                        if self.name:
+                            module_output_key = f"{self.name}.{key_name}"
+                            self.parent_module.foreach_modules[
+                                module_output_key
+                            ] = result
+                        return True
+        elif "[" in name:
+            tokens = name.split("[")
+            name = tokens[0]
+            if tokens[1] != "]":
+                num = tokens[1].replace("]", "")
+                if num.isdigit():
+                    index = int(num)
+                else:
+                    # Support Name[A], Name['A']
+                    index = num.strip('"').strip("'")
+
+        # Check if this is a Map parameter access with a specific key
+        if name not in self.parent_module.foreach_modules:
+            if name in self.props and isinstance(self.props[name], dict):
+                map_value = self.props[name]
+
+                # Handle !GetAtt MapName[Key] -
+                # return entire object at that key
+                if (
+                    index != -1
+                    and isinstance(index, str)
+                    and index in map_value
+                ):
+
+                    if prop_name == "":
+                        d[n] = map_value[index]
+                        return True
+
+                    # Handle !GetAtt MapName[Key].Attribute -
+                    # return specific attribute
+                    if (
+                        isinstance(map_value[index], dict)
+                        and prop_name in map_value[index]
+                    ):
+                        d[n] = map_value[index][prop_name]
+                        return True
+
+        return False
+
+    def find_ref(self, name):
+        """
+        Find a Ref.
+
+        A Ref might be to a module Parameter with a matching parent
+        template Property, or a Parameter Default. It could also
+        be a reference to another resource in this module.
+
+        :param name The name to search for
+        :return The referenced element or None
+        """
+        if name in self.props:
+            if name not in self.module_parameters:
+                # The parent tried to set a property that doesn't exist
+                # in the Parameters section of this module
+                msg = f"{name} not found in module Parameters: {self.source}"
+                raise exceptions.InvalidModuleError(msg=msg)
+
+            p = self.props[name]
+            if isdict(p):
+                if self.parent_module.name != "" and REF in p:
+                    p = self.parent_module.find_ref(p[REF])
+            return p
+
+        if name in self.module_parameters:
+            param = self.module_parameters[name]
+            if DEFAULT in param:
+                # Use the default value of the Parameter
+                return param[DEFAULT]
+            msg = f"{name} does not have a Default and is not a Property"
```

Comment:
I might be missing the point of having a Parameter with a Default for defining a list. Package doesn't support a parameter so it isn't changeable with that process?  Is this for sub-moduling?

```
Parameters:
  List:
    Type: CommaDelimitedList
    Default: A,B,C

Modules:
  Content:
    Source: ./foreach-module.yaml
    ForEach: !Ref List
    Properties:
      Name: !Sub my-bucket-$Identifier
```

If I can override it from a parent module or template then it would create some caveats on how the Outputs work. As in we can no longer guarantee that a particular value exists?

Reply:
For the first question, since this is all resolved client-side, we have to have either a default value for the parameter, or the parent template/module needs to specify it with a Property.

For the second, are you talking about conditional resources? Not sure how that will work with outputs that depend on a resource that is not actually emitted. Need to test it.